qualtran/_infra/data_types.py

```diff
@@ -354,6 +439,22 @@ def num_int(self) -> Union[int, sympy.Expr]:
     def fxp_dtype_str(self) -> str:
         return f'fxp-{"us"[self.signed]}{self.bitsize}/{self.num_frac}'
 
+    @property
+    def _fxp_dtype(self) -> Fxp:
+        return Fxp(None, dtype=self.fxp_dtype_str)
+
+    def to_bits(self, x: Union[float, Fxp]) -> Iterable[int]:
+        """Yields individual bits corresponding to binary representation of x"""
+        self.assert_valid_classical_val(x)
```

Comment:
This needs to be implemented still. Is it worth doing now?

Reply:
I think it would potentially change the logic in the next line. For example, if isinstance(x, Fxp) should we expect x to have the same _fxp_dtype (so no truncating occurs). For the float case I'm not really sure how one should handle it, perhaps we should only accept an Fxp float as input and not attempt to approximate it?

Reply:
Although, doing the "casting" from float to Fxp is very convenient...

Reply:
I've implemented the validation logic for `QIntOnesComp` and `QFxp`. However, implementing `assert_valid_classical_val` for `QFxp` opens a new can of worms because `Fxp` is currently not propagated correctly across the Bloqs; and is basically treated as a `QUint` in the classical simulation infrastructure; because of which the assertions fail in multiple bloq tests.

I don't want to open this can of worms as part of this PR. I have local updates to bloqs like `Join`, `Split` etc; all of which currently completely ignore `Fxp` and treat all classical data as unsigned integers; but we can do those updates as a follow-up PR

Reply:
That makes sense. Is this because of #722? I can prioritize if it's the case.